api/client/copy.go

```diff
@@ -0,0 +1,113 @@
+package client
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/docker/docker/pkg/archive"
+)
+
+type copyDirection int
+
+const (
+	fromContainer copyDirection = (1 << iota)
+	toContainer
+	acrossContainers = fromContainer | toContainer
+)
+
+// CmdCp handles copying files or directories to/from/across containers.
+func (cli *DockerCli) CmdCp(args ...string) error {
+	cmd := cli.Subcmd("cp", "[SRC_CONTAINER:]SRC_PATH [DST_CONTAINER:]DST_PATH", "Copy files/folders between containers or a local path")
+	if err := cmd.Parse(args); err != nil {
+		return nil
+	}
+
+	if cmd.NArg() != 2 {
+		cmd.Usage()
+		return nil
+	}
+
+	var (
+		srcPath      = cmd.Arg(0)
+		dstPath      = cmd.Arg(1)
+		srcContainer string
+		dstContainer string
+		direction    copyDirection
+	)
+
+	if strings.Contains(srcPath, ":") {
```

Comment:
I think this is currently an issue with the existing `docker cp` command and I have not yet figured out a way to solve it. I've kept windows paths in mind throughout most of this PR as I've opted to use `filepath.Separator` in lieu of `/` whenever possible, but I have no idea if this will work every time and I have not yet thought about how to deal with windows absolute paths which include a drive letter specifier like `C:\`